### ISSUE TITLE
torch.compiler.disable doesn't disable nested functions (also doesn't work as a context manager)

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#123771]
### 🐛 Describe the bug

```
import torch
torch.set_default_device('cuda')

@torch.compile(fullgraph=True)
def f(x):
    x = x.cos().cos()
    print(x)
    return x.cos()

def g(x):
    return f(x)

torch.compiler.disable(g)(torch.randn(5))
```
This still compiles (and throws the error).

Also, the docstring also claims it can be used as a context manager, but this error is thrown when you try to use it that way.

```
RuntimeError: torch._dynamo.optimize(...) is used with a context manager. Please refer to https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html to use torch._dynamo.optimize(...) as an annotation/decorator. 
```
cc: @anijain2305 

### Versions

N/A

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames @msaroufim @bdhirsh @anijain2305

**Comment by lukasschmit** [comment]
+1 using as a context manager would be very helpful!

**Comment by williamwen42** [comment]
I briefly looked into implementing this recently but ran into a two of design questions:

1. We should clarify the intended interaction between `compile` and `disable`. Do we really want `f` to not be compiled when `g` is called? One could argue that the `torch.compile` call could be viewed as an "enable" call, and in that case, what we observe is what is expected.

I have 2 different ideas on how `compile` and `disable` should interact:
- `compile` "enables" compilation and `disable` "disables" compilation. These two calls will override each other.
- `compile` "flags" compilation and `disable` "disables" compilation. `disable` will override `compile` but `compile` will not override `disable`. We can have another function `enable` that "enables" compilation - `disable` and `enable` will override each other.

2. Do we really need `disable` as a context manager? How do we envision using `disable` as a context manager? How would we get the disable context manager working in the case that we disable in the middle of a function?

e.g. should disable work in this example?
```
@torch.compile
def fn(...):
    x = x.cos().cos()
    with torch.compiler.disable():
        print(x)
        x = x + 1
    return x.cos()
```

Dynamo operates on frames, so disabling parts of a function would require more involved work. I'm thinking we could do this by:
- Graph breaking when tracing `disable`
- Create a new function from the bytecode contained in the context area (seems tricky to (1) determine which bytecode should extracted and (2) bytecode assemble the new function with the right globals, locals, etc.)
- Resume after the context

Of course, we could answer this question simply by removing the doc that says we can use `disable` as a context manager. I don't see any tests even using `disable` as a context manager anyway.

**Comment by anijain2305** [comment]
I don't think we need `enable`. `compile` flags/enables compilation. `disable` takes precedence over everything. If a frame is marked `disable`, anything underneath should not be compiled. 

We don't want context manager. TorchDynamo was originally a context manager, and it caused more issues than solutions. The point about context manager is mostly about why the codepath is even going there.

**Comment by ezyang** [comment]
Setting aside context manager, I still think we need to be more precise when we are talking about potential semantics for compile/disable. For me, it is most helpful when I think about what the internal implementation of torch.compile is: it installs a dynamically scoped eval frame handler, which causes us to start attempting to compile frames every time we enter a frame.

From the perspective of an "eval frame handler" view, it makes perfect sense that disable around compile does nothing, because what is disable even going to do? It can't reset the frame handler to the basic one, because compile will go ahead and reset it back to the compiling handler when it rolls around. So we must introduce some new concept to get the semantics that Horace wants.

The simplest implementation is to have a boolean flag "compile enabled", orthogonal from the frame handler. You now have:

* "compile enabled" is true by default
* torch.compile - install frame handler, leave "compile enabled" as is
* torch.compiler.disable - uninstall frame handler, set "compile enabled" false

Looking over the state transitions here, it becomes clear that an "enable" function is needed, because once you set "compile enabled" to false there is no way to get out of it.

I'm not entirely sure this is going to be that understandable to users though, I fear...

**Comment by ezyang** [comment]
Also, I think https://github.com/pytorch/pytorch/issues/131263 is a bit different. Under the compile enabled boolean flag proposal, enable around disable still nets out as disabled, so you need something *else* to mean "this *particular* function was marked as disabled, and now I'm going to make that annotation inert".

**Comment by Chillee** [comment]
@ezyang Perhaps there needs to be a more thought out system around enabling and disabling parts of compilation 🤔 Or perhaps just a context manager for enable and disable is sufficient.

**Comment by ezyang** [comment]
I guess we need to (sigh) gather user requirements lol

**Comment by Chillee** [comment]
@ezyang I think my first take would be that "a context manager for enable and disable" is sufficient.

**Comment by ezyang** [comment]
OK marking this as actionable

**Comment by williamwen42** [comment]
I started making some progress but I need to address the following 2 design issues:
1. Stacked decorators

It's not completely clear how the dynamo decorators should behave when stacked. Currently (before my PRs), the only the outermost decorator is applied, e.g.

```python
import torch

def be1(gm, ex):
    print("backend 1")
    return gm

def be2(gm, ex):
    print("backend 2")

@torch.compile(backend=be1)
@torch.compile(backend=be2)
def fn(x):
    return x + 1

fn(torch.randn(3, 3))
```

outputs `backend 1`.

But typically, the innermost decorator is the one that is applied, e.g.

```python
import torch

@torch.no_grad
@torch.enable_grad
def f():
    print(torch.is_grad_enabled())

f()
```

outputs `True`.

Should we change the dynamo decorators to only apply the innermost compile/disable/enable?

2. Context managers

For example,
```python
@torch.compile
def f(x):
    x = x + 1
    with torch._dynamo.disable():
        x = x + 1
        x = f2(x)
    return x

@torch._dynamo.disable
def g(x):
    with torch._dynamo.enable():
        x = x + 1
        x = g2(x)
    return x

@torch.compile
def h(x):
    return g(x)
```

It seems to me it would be easiest for the decorators to enable/disable compile on function calls made in the context manager region, NOT the bytecode in the context manager region itself. i.e. in `f`, `x = x + 1` would still be traced, but `f2(x)` won't. In `g`, `x = x + 1` would not be traced, but `g2(x)` would.

**Comment by anijain2305** [comment]
For 1), yes, you are right. Inner most one should be applied. I don't remember know if there are some repercussions of inverting the behavior now.

For 2), this confusion is one of the reasons, we disabled the context manager usage of torch.compile. I still think we should not take the context manager route (unless I am misunderstanding the user intent here).

**Comment by ezyang** [comment]
1) You can't actually invert it, UNLESS you are willing to make a special case when a torch.compile goes directly to another torch.compile, but that's whack.

The reason you can't invert is because of this:

```
@torch.compile(backend="foo")
def f(x):
  ...

@torch.compile(backend="bar")
def g(x):
  return f(x)
```

In this program, clearly g should compile with backend bar, and inline the contents of f in. No other semantics makes sense.

2) For something that is only toggling the "disable" bit, a context manager would be OK. But if you're also mucking with the frame handler, which I think folks have argued is more intuitive, you are also forced to do decorator only.

**Comment by williamwen42** [comment]
1. I was only thinking of the whack case of multiple compile decorators in a row. But this might no longer be an issue since I managed to the enable/disable implementation from the compile context manager.

**Comment by williamwen42** [comment]
@Chillee @ezyang during implementation I discovered that HOPs actually depend on the current disable/compile interaction where compile overrides disable (see test failures in https://github.com/pytorch/pytorch/pull/132925). So this change could potentially impact users who depend on this interaction. If we go through with disable overriding compile in all cases, then enable in the HOP case should fix things.

**Comment by anijain2305** [comment]
Cc @ydwu4

**Comment by mlazos** [comment]
FWIW, I think the current semantics make sense. I view disable as a compiler directive, and if it's run outside of the compiler it means nothing. However since I'm late and it looks like we've started on implementing this I'm okay if we've established we want to change the semantics. In light of the BC breaking piece that William has encountered perhaps we should reconsider?

I think Ed's proposal of having an enable and disable would make the most intuitive sense to users and allows us the flexibility of having an answer for users who want the existing behavior (they can just add an enable somewhere) even though it breaks BC.

**Comment by Chillee** [comment]
@williamwen42 Yeah I think we should specifically enable `torch.compile` inside the HOPs.

**Comment by ydwu4** [comment]
Yeah, @williamwen42 as we discussed today, 1. the nested control flow op case would require re-enable the disabled function 2. make_fx a higher order op would also require re-enable the disabled region. 

The simplest solution would be annotating the the wrapper function in each compiled control flow operator with enable.

**Comment by williamwen42** [comment]
@Chillee @lukasschmit @anijain2305 @ezyang trying to gather requirements for the context manager. We used to support a context manager for `torch.compile` but we removed that functionality due to confusing behavior. There is still pushback on reintroducing context managers for `enable/disable` due to potentially confusing behavior.

Consider the following example:

```python
def a(x):
    return x + 1

def b(x):
    torch._dynamo.graph_break()
    return x + 1

@compile
def c(x):
    with disable():
        x = b(x)
        x = torch.sin(x)
    return x

@compile
@disable
def d(x):
    with enable():
        x = a(x)
        x = torch.cos(x)
    return x
```

There are 2 options that I can think of for context manager behavior:
1. `enable/disable` only apply to traced (i.e. not in-graph) function calls in the context manager - ops are not affected by the context manager. In the example, when running `c`, `b` is skipped (graph break on attempting to call `b`) but `torch.sin` is traced in the graph. When running `d`, `a` is traced but `torch.cos` is skipped (no additional graph break). This behavior is easier to implement (because Dynamo takes frames as input, not bytecode sections), but is potentially less intuitive (since in-graph operators will ignore the context manager). 
2. `enable/disable` apply to all instructions in the context manager. In the example, when running `c`, `b` and `torch.sin` are skipped (1 graph break). When running `d`, `a` and `torch.cos` are traced. This behavior is harder to implement (we would probably have to separate the context managed region into its own function), but it is more intuitive.

**Comment by ezyang** [comment]
Let's work out the BC problem in the PR first I think

**Comment by williamwen42** [comment]
Functionality implemented by `torch.compiler.set_stance`, merged in https://github.com/pytorch/pytorch/pull/137504

**Comment by Edenzzzz** [comment]
@williamwen42 It seems that `torch.compiler.set_stance` is not available in the newest PyTorch (2.5.1). Currently I still don't see a way to force eager excecution of a function decorated with `torch.compile` from an outer function, as mentioned in [`torch.compiler.force_uncompiled`](https://github.com/pytorch/pytorch/pull/132926#issuecomment-2375777245). The `set_stance` [PR](https://github.com/pytorch/pytorch/pull/137504) is also closed?

**Comment by Edenzzzz** [comment]
Nvm, the nightly version has it.
